core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -682,6 +683,17 @@ private[log] class Cleaner(val id: Int,
         else
           Cleaner.this.shouldRetainRecord(map, retainDeletesAndTxnMarkers, batch, record, stats)
       }
+
+      override def retrieveDeleteHorizon(batch: RecordBatch) : Long = {
+        if (batch.isDeleteHorizonSet())
+          return batch.deleteHorizonMs() // means that we keep the old timestamp stored
+
+        // check that the control batch has been emptied of records
+        // if not, then we do not set a delete horizon until that is true
+        if (batch.isControlBatch() && transactionMetadata.onControlBatchRead(batch))
```

Comment:
Hmm, we already called transactionMetadata.onControlBatchRead(batch) in shouldDiscardBatch(). So, we probably don't want to call it again on the same batch. Instead, we want to pass along the return value to retrieveDeleteHorizon().

Reply:
Yep, working on this. Will address this soon.